### PR TITLE
Attempt to lock down dependencies

### PR DESCRIPTION
**Pull request by relrod**
In the current world, we do not lock down dependencies at all. This means that developers are all potentially developing against moving targets, and some dependencies from which we subclass models could generate migrations (which could be backwards incompatible and cause issues for environments that use system packages pinned to older versions).

In this commit is really 3 changes:

- Update the updater script a bit - let it write out all the DAB application .txt files
- Make pyproject.toml use those .txt files
- Add current working/known upper bounds to all deps where possible before generating the .txt files.

I believe this is the first step towards sanity here and vision into which versions of dependencies are getting pulled in and where.

In the current world, we have a dep that released a new z-stream release which changed a field in a baseclass we're using. For developers who got that new version of the library (since we don't pin deps currently), it is easy to think that we need to push the new migration that gets generated. However we have no idea currently if that change would be compatible with apps using DAB, or apps forced to use the old version of the dependency due to system packages.

### CODE REVIEW

File: requirements/requirements.in [hunk 1]
```diff
@@ -2,9 +2,9 @@
...
-Django>=4.2.5,<4.3.0
-djangorestframework
-django-crum
-django-split-settings
-inflection
+cryptography<=42.0.5
+Django>=4.2.5,<4.2.15
+djangorestframework<=3.15.1
+django-crum<=0.7.9
+django-split-settings<=1.2.0
```
Comment (AlanCoding) [inline requirements/requirements.in:9]: I guess it's a personal preference thing, but I don't feel like this needs to be so opinionated. That's what the updater script gets us.

For this, a pin of `<1.3` or `<2` might be reasonable. Not upgrading to the next _major_ version is probably generally reasonable. But not even that's necessary. You could just leave everything unpinned until there's an issue and comment what the problem is.

File: requirements/requirements_authentication.in [hunk 2]
```diff
@@ -1,17 +1,17 @@
-social-auth-app-django
-tabulate
+social-auth-app-django<=5.4.1
```
Comment (AlanCoding) [inline requirements/requirements_authentication.in:1]: This is duplicated with, and matches, `requirements_all.txt`. I'm not sure if that's theoretically guaranteed (resolving dependencies in isolation, vs with other files), but if it's not I feel like that would suggest other contradictions which are a problem anyway.

**Review by AlanCoding — APPROVED** [review]
I see more risk from not doing than from doing it. Shipping with floating dependencies is bad for us. This could prevent a future mess where we have to go back and fill these in via a semi-manual process. Instead of thinking through that, I'd rather just hit the merge button now.

File: requirements/requirements.in (hunk 1 above)
Comment (relrod) [reply requirements/requirements.in:9]: @AlanCoding One of two separate things that sent me down this rabbit hole in the first place was that `social-auth-app-django` released a z-version (5.4.2) which changed a field and thus triggers a migration in DAB[authentication].

We have no idea if that migration would break a system package that pins to 5.4.1.

File: requirements/requirements_authentication.in (hunk 2 above)
Comment (relrod) [reply requirements/requirements_authentication.in:1]: The app-specific .txt files get generated with passing `-c requirements_all.txt` to `pip`. Basically it generates the _all.txt file and then uses that for consistency in generating the app-specific .txt files.

Comment (AlanCoding) [reply requirements/requirements_authentication.in:1]: amazing.

File: requirements/requirements.in (hunk 1 above)
Comment (john-westcott-iv) [inline requirements/requirements.in:9]: Do we really want to pin the in files like this? We typically only pin major requirements (like django) or if there is a known issue with a version. Since we pointed everything to requirements_*.txt I feel like these should be opened unless there is an issue.

File: requirements/requirements_dev.txt [hunk 3]
```diff
@@ -2,9 +2,9 @@ ansible  # Used in build process to generate some configs
 build
 Django>=4.2.5,<4.3.0
 django-debug-toolbar
-django-extensions
-djangorestframework
-django-split-settings
+django-extensions<=3.2.3
+djangorestframework<=3.15.1
+django-split-settings<=1.2.0
```
Comment (john-westcott-iv) [inline requirements/requirements_dev.txt:7]: Are these causing issues? Or did we just pin them because we pinned them in other places?

File: requirements/updater.sh [hunk 4]
```diff
@@ -1,62 +1,69 @@
...
 PYTHON=python3.11
 
 for FILE in requirements.in requirements_all.txt ; do
-	if [ ! -f ${FILE} ] ; then
-		touch ${FILE}
-	fi
+  if [ ! -f ${FILE} ] ; then
+    touch ${FILE}
+  fi
 done
```
Comment (john-westcott-iv) [inline requirements/updater.sh:12]: I'm wondering if we should also parse pyproject.toml and create empty files for all the specified dependencies in there.

File: requirements/updater.sh [hunk 5]
```diff
@@ -1,62 +1,69 @@
...
-  ${venv}/bin/python -m pip install -U 'pip' pip-tools
+  ${venv}/bin/pip install -U pip pip-tools
+
+  ${pip_compile} ${requirements_dir}/requirements*.in --output-file requirements_all.txt
 
-  ${pip_compile} $(ls ${requirements_dir}/requirements*.in | xargs) --output-file requirements_all.txt
+  for file in ${requirements_dir}/requirements*.in; do
+    app="$(basename $file .in)"
+    echo "Compiling deps for $app"
+    ${pip_compile} $file --output-file "$app.txt" -c requirements_all.txt
```
Comment (john-westcott-iv) [inline requirements/updater.sh:35]: The -c is just protecting you here incase there is a release of some library between when you created requirements_all.txt and these individual files right? Or is it doing something more?

File: requirements/updater.sh [hunk 6]
```diff
@@ -1,62 +1,69 @@
...
+  done
 }
 
 main() {
   base_dir=$(pwd)
 
-  _tmp=$(${PYTHON} -c "import tempfile; print(tempfile.mkdtemp(suffix='.ansible_base-requirements', dir='/tmp'))")
+  _tmp=$(mktemp -d /tmp/tmp.XXXXXX.ansible_base-requirements)
```
Comment (john-westcott-iv) [inline requirements/updater.sh]: We did the python thing here because of subtle differences between mktmp on OS X and Fedora (and potentally other distros). For example, this mktmp command on OS X gives the following literal directory:
```
/tmp/tmp.XXXXXX.ansible_base-requirements
```

Whereas the same command on Fedora will actually put a value in there:
```
/tmp/tmp.3AqBMX.ansible_base-requirements
```

File: test_app/tests/lib/redis/test_client.py [hunk 7]
```diff
@@ -312,7 +312,7 @@ def test_get_redisSTATUS_FAILED_redis_mode():
 @mock.patch('ansible_base.lib.redis.client.DABRedisCluster.cluster_nodes', return_value='there')
 # Mock the RedisCluster functions that attempt to connect on definition
 @mock.patch('redis.cluster.NodesManager.initialize')
-@mock.patch('redis._parsers.CommandsParser.initialize')
+@mock.patch('redis.commands.CommandsParser.initialize')
```
Comment (john-westcott-iv) [inline test_app/tests/lib/redis/test_client.py:315]: Is this accidentally added? It seems unrelated.

File: requirements/updater.sh (hunk 5 above)
Comment (relrod) [reply requirements/updater.sh:35]: -c provides consistency across _all_ the deps, rather than resolving the deps for each app independently. Otherwise you end up with some apps trying to pull in Django 5 for example.

File: test_app/tests/lib/redis/test_client.py (hunk 7 above)
Comment (relrod) [reply test_app/tests/lib/redis/test_client.py:315]: It is not unrelated, and this confusion shows the exact reason for this PR existing (luckily in this case the problem only manifested itself in a test file). We're developing against different versions than the downstream RPMs are targeting. In this case, the class was moved to a different module in the newer version of Redis.

File: requirements/updater.sh (hunk 6 above)
Comment (relrod) [reply requirements/updater.sh]: Yeah on BSD/MacOS, the template has to come at the end. So `mktemp -d /tmp/ansible_base-requirements.XXXXXX` should work.


```suggestion
  _tmp=$(mktemp -d /tmp/ansible_base-requirements.XXXXXX)
```

**Review by shanemcd — APPROVED** [review]
This bit us recently with @newswangerd's legacy SSO code.

**Comment by shanemcd** [comment]
Thank you for doing this @relrod .
